### ISSUE TITLE
JTAG: can't read PC, can't step, can't attach gdb

### ISSUE THREAD

**Issue by Strobokopp** [lambdaconcept/minerva#8]
I'm missing some basics here for sure, but maybe you can help me find out what I can expect and what to look at.

Situation: I have a small design with minerva, a Wishbone bus that is shared between cpu.ibus and cpu.dbus through a priority arbiter, and some devices on the bus:
- ROM (nmigen-generated) with some bare-metal instructions at the reset vector
- some bytes of SRAM (nmigen-generated)
- a controller that interfaces a 128 MBit NOR flash
- 8 LEDs I can write to and read from at one single address..

JTAG issues, using OpenOCD:
```
Open On-Chip Debugger 0.10.0+dev-00626-g7eaf60f1b (2019-07-29-16:00)
[...]
Info : BCM2835 GPIO JTAG/SWD bitbang driver
Info : JTAG only mode enabled (specify swclk and swdio gpio to add SWD mode)
Info : RCLK (adaptive clock speed) not supported - fallback to 200 kHz
Info : JTAG tap: minerva.cpu tap/device found: 0x10e31913 (mfg: 0x489 (SiFive, Inc.), part: 0x0e31, ver: 0x1)         
Info : datacount=1 progbufsize=0       
Warn : We won't be able to execute fence instructions on this target. Memory ma
y not always appear consistent. (progbufsize=0, impebreak=0)
Info : Examined RISC-V core; found 32 harts
Info :  hart 0: XLEN=32, misa=0x0
[...]
```
- after doing `halt` I can read the registers x1,x2,x3... but reading x32 (PC) always returns 0.
- with the CPU in halted state, doing `step` fails, though afterwards I can `resume` manually:
```
> step
unable to resume hart 0
  dmstatus =0x00020882
  was stepping, halting
```
- GDB can't attach to my OpenOCD server:
```
GNU gdb (GDB) 8.3.0.20190516-git
[...]
(gdb) target remote rpi:3333
Remote debugging using rpi:3333
bfd requires xlen 8, but target has xlen 4
```
Any help understanding these issues is greatly appreciated.

**Comment by jfng** [comment]
Sorry for the late response. I was able reproduce the behaviour you described. These are bugs and they should be fixed in 53251ba. Feel free to re-open this issue if this is not the case !

**Comment by Strobokopp** [comment]
I can confirm that the first two issues (reading PC, stepping) are resolved. Great! :)
For connecting to the core with GDB I found out that I have to explicitly set the arch before using the `target` command. Otherwise I still get the `bfd requires xlen 8, but target has xlen 4` message.
```
(gdb) set arch riscv:rv32
The target architecture is assumed to be riscv:rv32
(gdb) target remote rpi:3333
Remote debugging using rpi:3333
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
(gdb)
```
So that works. I couldn't make out if this is a bug or not. GDB requests information about the target from OpenOCD and I would expect that to include whether it's RV32 or RV64. I attached a riscv.xml file that OpenOCD wrote after the `gdb_save_tdesc` command. As far as I understood that's exactly the information GDB receives.
Also attaching my OpenOCD config for minerva.
[openocdfiles.zip](https://github.com/lambdaconcept/minerva/files/4605404/openocdfiles.zip)
